aten/src/ATen/native/cudnn/RNN.cpp

```diff
@@ -0,0 +1,896 @@
+    weight_buf.zero_();
+    std::vector<Tensor> params;
+    size_t params_stride0;
+    std::tie(params, params_stride0) = get_parameters(handle, fn.rnn, descs.rnn_desc, descs.x_descs[0], w_desc, weight_buf);
+    _copyParams(MatrixRef<Tensor>{weight, static_cast<size_t>(weight_stride0)},
+                MatrixRef<Tensor>{params, params_stride0});
+  } else {
+    w_desc.set(weight_buf, 3);
+  }
+
+  if (cx.defined() && !cx.sizes().equals(hidden_size)) {
+    std::ostringstream oss;
+    oss << "Expected cell size " << IntList{hidden_size} << ", got " << cx.sizes();
+    throw std::runtime_error(oss.str());
+  }
+
+  size_t workspace_size;
+  auto x_descs_arr = descs.get_x_descs();
+  auto y_descs_arr = descs.get_y_descs();
+  CUDNN_CHECK(cudnnGetRNNWorkspaceSize(
+        handle,
+        descs.rnn_desc.desc(),
+        fn.tensors.seq_length,
+        x_descs_arr.data(),
+        &workspace_size
+        ));
+  Tensor workspace = input.type().toScalarType(kByte).tensor(workspace_size);
+
+  Tensor reserve;
+  // NB: Previously, the test was for fn.requires_grad, but we don't have
+  // this information.  Use 'train' as a proxy.
+  if (fn.dropout.train) {
+    size_t reserve_size;
+    CUDNN_CHECK(cudnnGetRNNTrainingReserveSize(
+          handle,
+          descs.rnn_desc.desc(),
+          fn.tensors.seq_length,
+          x_descs_arr.data(),
+          &reserve_size
+          ));
+    reserve = input.type().toScalarType(kByte).tensor(reserve_size);
+    CUDNN_CHECK(cudnnRNNForwardTraining(
+          handle,
+          descs.rnn_desc.desc(),
+          fn.tensors.seq_length,
+          x_descs_arr.data(), x.data_ptr(),
+          descs.hx_desc.desc(), hx.data_ptr(),
+          descs.cx_desc.desc(), cx.defined() ? cx.data_ptr() : nullptr,
+          w_desc.desc(), weight_buf.data_ptr(),
+          y_descs_arr.data(), y.data_ptr(),
+          descs.hy_desc.desc(), hy.data_ptr(),
+          descs.cy_desc.desc(), cy.defined() ? cy.data_ptr() : nullptr,
+          workspace.data_ptr(), workspace.size(0),
+          reserve.data_ptr(), reserve.size(0)
+          ));
+  } else { // inference
+    reserve = input.type().toScalarType(kByte).tensor();
+    CUDNN_CHECK(cudnnRNNForwardInference(
+          handle,
+          descs.rnn_desc.desc(),
+          fn.tensors.seq_length,
+          x_descs_arr.data(), x.data_ptr(),
+          descs.hx_desc.desc(), hx.data_ptr(),
+          descs.cx_desc.desc(), cx.defined() ? cx.data_ptr() : nullptr,
+          w_desc.desc(), weight_buf.data_ptr(),
+          y_descs_arr.data(), y.data_ptr(),
+          descs.hy_desc.desc(), hy.data_ptr(),
+          descs.cy_desc.desc(), cy.defined() ? cy.data_ptr() : nullptr,
+          workspace.data_ptr(), workspace.size(0)
+          ));
+
+  }
+
+  if (batch_first && !is_input_packed) {
+    output.transpose_(0, 1);
+  }
+
+  return std::make_tuple(output, hy, cy, reserve, weight_buf);
+}
+
+std::tuple<Tensor, Tensor, Tensor> _cudnn_rnn_backward_grad(
```

Comment:
`backward_grad` doesn't really tell you what it is (both terms are kind of interchangeable and you also compute grad of weights). How about `backward_input`?